src/fragmentation_manager.js

```diff
@@ -30,115 +13,57 @@ class FragmentationManager {
+		// * Some extra fragment
+		if (packet.fragmentId !== 0) {
+			this.fragments.push({
+				sequenceId: packet.sequenceId,
+				fragmentId: packet.fragmentId,
+				nextedExpectedSequenceId: packet.sequenceId+1,
+				payload: packet.payload
+			});
 		}
 
-		const payload = packet.payload;
-		const uninqueSequenceAndConnectionId = this.makeUniqueId(packet.sequenceId);
+		// * End of fragmentation. Start rebuilding
+		if (packet.fragmentId === 0) {
+			const fragments = [];
+			const secondToLastFragment = this.fragments.find(fragment => fragment.nextedExpectedSequenceId === packet.sequenceId);
 
-		this.seenIds.push(uninqueSequenceAndConnectionId);
-
-		if (!this.fragments[uninqueSequenceAndConnectionId]) {
-			this.fragments[uninqueSequenceAndConnectionId] = {
-				id: packet.fragmentId,
-				payload: payload
-			};
-		}
-
-		let { defragmented } = this.fragments[uninqueSequenceAndConnectionId];
-
-		if (!defragmented) {
-			defragmented = [];
-
-			if (packet.fragmentId === 0) {
+			if (secondToLastFragment) {
+				// * Payload was fragmented
 				const missingFragments = [];
-				for (let i = packet.sequenceId - 1; i >= Math.max(packet.sequenceId - 50, this.startSequenceIds[this.baseId]); i--) {
-					const id = this.makeUniqueId(i);
-					if (!this.seenIds.includes(id)) {
-						missingFragments.push(id);
-						//console.log('Missing packet', id, 'in stream, deferrring fragment restoration');
-					}
-				}
-
-				for (const id of missingFragments) {
-					this.deferredFragments[id] = {
-						missingFragments: missingFragments,
-						sequenceId: packet.sequenceId
-					};
-				}
-
-				if (missingFragments.length === 0) { // nothing missing, restore the fragments
-					let previousFragment = this.fragments[this.makeUniqueId(packet.sequenceId - 1)];
-					if (previousFragment && previousFragment.id > packet.fragmentId) {
-						console.log('Restoring', previousFragment.id, 'fragments');
-						for (let i = packet.sequenceId; i >= this.startSequenceIds[this.baseId] - 1; i--) {
-							let fragment = this.fragments[this.makeUniqueId(i)];
-							if (!fragment) {
-								console.error('Cannot find fragment', i, 'in the past for packet', uninqueSequenceAndConnectionId);
-								break;
-							}
-
-							defragmented.push(fragment.payload);
-
-							if (fragment.id == 1) {
-								break;
-							}
-						}
-					} else {
-						// no additional fragments detected
-						defragmented.push(payload);
-					}
-				}
-			} else { // missing fragments, attempt to defer defragmentation until a later stage
-				const deferredFragment = this.deferredFragments[uninqueSequenceAndConnectionId];
-				if (deferredFragment) {
-					console.log('Found missing packet', uninqueSequenceAndConnectionId, ' fragment: ', packet.fragmentId, ')');
-					const { missingFragments } = deferredFragment;
+				const firstFragment = this.fragments.find(fragment => fragment.sequenceId === secondToLastFragment.sequenceId-secondToLastFragment.fragmentId+1);
```

Comment:
Here we are assuming that we will find the first fragment, when the fragment could be missing.

Also, we don't seem to keep track of the sequence IDs for missing packets anymore?

Reply:
The missing fragment sequence IDs do get "tracked", further on

https://github.com/PretendoNetwork/nex-viewer/blob/cc7d8b9910e1d59fc85e7cb2b3bbdce2d4cdeeff/src/fragmentation_manager.js#L44

Nothing is done with them though since I don't have a dump which has missing fragments, which makes it difficult to test deferred fragmentation handling. This is also why the first fragment is always assumed to exist

I made this PR using a new branch here though, so you (or anyone else) could pull it if you have dumps with missing fragments. It seems to happen fairly rarely, so I can't reliably get dumps like that myself

Reply:
Sounds good! I've been testing the PR with no issues, so go ahead